staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go

```diff
@@ -0,0 +1,178 @@
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugins
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/klog/v2"
+
+	v1 "k8s.io/api/core/v1"
+	storage "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// VSphereDriverName is the name of the CSI driver for vSphere Volume
+	VSphereDriverName = "csi.vsphere.vmware.com"
+	// VSphereInTreePluginName is the name of the in-tree plugin for vSphere Volume
+	VSphereInTreePluginName = "kubernetes.io/vsphere-volume"
+	// Name of parameter to supply SPBM Policy name for Volume Provisioning
+	paramStoragePolicyName = "storagepolicyname"
+	// Name of parameter to let controller know that it needs to return volumefilePath as Volume Attribute
+	// When CSI Driver Provision Volume using in-tree Storage Class and PVC
+	paramRequestVolumeFilePath   = "requestvolumefilepath"
+	paramVolumeFilePathAttribute = "volumefilepath"
+)
+
+var _ InTreePlugin = &vSphereCSITranslator{}
+
+// vSphereCSITranslator handles translation of PV spec from In-tree vSphere Volume to vSphere CSI
+type vSphereCSITranslator struct{}
+
+// NewvSphereCSITranslator returns a new instance of vSphereCSITranslator
+func NewvSphereCSITranslator() InTreePlugin {
+	return &vSphereCSITranslator{}
+}
+
+// TranslateInTreeStorageClassToCSI translates InTree vSphere storage class parameters to CSI storage class
+func (t *vSphereCSITranslator) TranslateInTreeStorageClassToCSI(sc *storage.StorageClass) (*storage.StorageClass, error) {
+	if sc == nil {
+		return nil, fmt.Errorf("sc is nil")
+	}
+	var params = map[string]string{}
+	// only "storagepolicyname" SC parameter in vsphere in-tree volume is
+	// supported by vSphere CSI Driver.
+	for k, v := range sc.Parameters {
+		paramName := strings.ToLower(k)
+		if paramName == paramStoragePolicyName {
+			params[paramStoragePolicyName] = v
+		} else if paramName == fsTypeKey {
+			// translate deprecated storage class parameter key fstype to csi.storage.k8s.io/fstype
+			params[csiFsTypeKey] = v
+		} else {
+			klog.V(2).Infof("StorageClass parameter [name:%q, value:%q] is not supported by the vSphere CSI driver. "+
+				"Translated Storage Class will not have this parameter.", paramName, v)
+		}
+	}
+	// sc.AllowedTopologies for Topology based volume provisioning will be supplied as it is.
+
+	// requestvolumefilepath param will help CSI Controller identify whether it needs to make a call to
+	// fetch vmdk path after Volume is created.
+	params[paramRequestVolumeFilePath] = "true"
```

Comment:
As discussed, lets remove this and add a `csiMigrationParams`.

Reply:
adding `csimigrationparams` which will hold JSON encoded string for all in-tree parameter keys and values.

```
// CSIMigrationParams holds all in-tree vSphere volume Parameters Key and values
// in the json encoded string
CSIMigrationParams = "csimigrationparams"
```